src/python/pants/option/native_options.py

```diff
@@ -0,0 +1,140 @@
+# Copyright 2024 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import inspect
+import logging
+import shlex
+from enum import Enum
+from typing import Any, Mapping, Optional, Sequence, Tuple
+
+from pants.engine.internals import native_engine
+from pants.engine.internals.native_engine import PyConfigSource
+from pants.option.config import ConfigSource
+from pants.option.custom_types import _flatten_shlexed_list, shell_str
+from pants.option.errors import OptionsError
+from pants.option.ranked_value import Rank
+from pants.util.strutil import get_strict_env
+
+logger = logging.getLogger()
+
+
+class NativeOptionParser:
+    """A Python wrapper around the Rust options parser."""
+
+    int_to_rank = [
+        Rank.NONE,
+        Rank.HARDCODED,
+        Rank.CONFIG_DEFAULT,
+        Rank.CONFIG,
+        Rank.ENVIRONMENT,
+        Rank.FLAG,
+    ]
+
+    def __init__(
+        self,
+        args: Optional[Sequence[str]],
+        env: Mapping[str, str],
+        config_sources: Optional[Sequence[ConfigSource]],
+        allow_pantsrc: bool,
+    ):
+        py_config_sources = (
+            None
+            if config_sources is None
+            else [PyConfigSource(cs.path, cs.content) for cs in config_sources]
+        )
+        self._native_parser = native_engine.PyOptionParser(
+            args,
+            dict(get_strict_env(env, logger)),
+            py_config_sources,
+            allow_pantsrc,
+        )
+
+        # (type, member_type) -> native get for that type.
+        self._getter_by_type = {
+            (bool, None): self._native_parser.get_bool,
+            (int, None): self._native_parser.get_int,
+            (float, None): self._native_parser.get_float,
+            (str, None): self._native_parser.get_string,
+            (list, bool): self._native_parser.get_bool_list,
+            (list, int): self._native_parser.get_int_list,
+            (list, float): self._native_parser.get_float_list,
+            (list, str): self._native_parser.get_string_list,
+            (dict, None): self._native_parser.get_dict,
+        }
+
+    def get(
+        self, *, scope, flags, default, option_type, member_type=None, passthrough=False
+    ) -> Tuple[Any, Rank]:
+        def is_enum(typ):
+            # TODO: When we switch to Python 3.11, use: return isinstance(typ, EnumType)
+            return inspect.isclass(typ) and issubclass(typ, Enum)
+
+        # '--foo.bar-baz' -> ['foo', 'bar', 'baz']
+        name_parts = flags[-1][2:].replace(".", "-").split("-")
+        switch = flags[0][1:] if len(flags) > 1 else None  # '-d' -> 'd'
+        option_id = native_engine.PyOptionId(*name_parts, scope=scope or "GLOBAL", switch=switch)
+
+        rust_option_type = option_type
+        rust_member_type = member_type
+
+        if option_type is dict:
```

Comment:
Below is a bunch of munging to massage the raw rust values into their expected Python types.

Reply:
It looks like there's some pretty... intricate processing here with the `eval` calls and the `shlex.split`ing.

Might we ever end up with fully Rust-side manipulation of options, or would we expecting to always go back out through Python for these sort of adjustments? If the first one, is there a path for moving more of this logic into Rust too?

Reply:
Yeah, this mostly emulates what happens in the native parser, which unfortunately supports all sorts of clever types, expressed as a str -> type cast.  

I don't yet know which of these will make sense to represent in Rust. At a guess, I'd say shell_str and path yes, but targets no?

Anyway, I think this will be best revisited after we get rid of the legacy parser.  Then we can do more chipping away at the Python side of things once a whole ton of that cruft is gone.